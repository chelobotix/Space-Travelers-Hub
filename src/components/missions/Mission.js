import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import style from './Mission.module.css';
import { joinMission, leaveMission } from '../../redux/features/missions/missionsSlice';

const Mission = (props) => {
  const dispatch = useDispatch();
  const {
    // eslint-disable-next-line camelcase
    id, name, description, reserved,
  } = props;

  const handleJoinMission = (id) => {
    dispatch(joinMission(id));
  };

  const handleLeaveMission = (id) => {
    dispatch(leaveMission(id));
  };

  return (
    <>
      <td data-testid="titleTD" className={`${style.border} ${style.title}`}>
        {name}
      </td>
      <td className={`${style.border} ${style.description}`}>
        {description}
      </td>
      <td className={`${style.border} ${style.status}`}>
        {!reserved
          ? <span className={style.status1}>NOT A MEMBER</span>
          : <span className={style.status2}>Active Member</span>}
      </td>
      <td className={`${style.border} ${style.status}`}>
        {!reserved
          ? <button className={style.btn1} onClick={() => handleJoinMission(id)} type="button">Join Mission</button>
          : <button className={style.btn2} onClick={() => handleLeaveMission(id)} type="button">Leave Mission</button>}
      </td>
    </>
  );
};

Mission.defaultProps = {
  id: '', name: '', description: '', reserved: false,
};
Mission.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  description: PropTypes.string,
  reserved: PropTypes.bool,
};

export default Mission;
