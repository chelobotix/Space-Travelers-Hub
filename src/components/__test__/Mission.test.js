import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { render, screen, fireEvent } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import Mission from '../missions/Mission';

const missionsArray = [
  {
    mission_id: '1',
    mission_name: 'Mission test',
    description: 'Description Test',
    reserved: false,
  },
];

const mockStore = configureStore([]);
const initialState = {
  missions: {
    missions: missionsArray,
  },
};

const store = mockStore(initialState);

describe('Mission', () => {
  test('Get Mission Title', () => {
    render(
      <Provider store={store}>
        <table>
          <tbody>
            <tr>
              <Mission
                id={missionsArray[0].mission_id}
                name={missionsArray[0].mission_name}
                description={missionsArray[0].description}
                reserved={missionsArray[0].reserved}
              />
            </tr>
          </tbody>
        </table>
      </Provider>,
    );
    const td = screen.getByTestId('titleTD');
    expect(td).toHaveTextContent('Mission test');
  });

  test('Mission Tree', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <table>
          <tbody>
            <tr>
              <Mission
                id={missionsArray[0].mission_id}
                name={missionsArray[0].mission_name}
                description={missionsArray[0].description}
                reserved={missionsArray[0].reserved}
              />
            </tr>
          </tbody>
        </table>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Click button text', () => {
    render(
      <Provider store={store}>
        <table>
          <tbody>
            <tr>
              <Mission
                id={missionsArray[0].mission_id}
                name={missionsArray[0].mission_name}
                description={missionsArray[0].description}
                reserved={missionsArray[0].reserved}
              />
            </tr>
          </tbody>
        </table>
      </Provider>,
    );
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(button.textContent).toBe('Join Mission');
  });

  test('Click button listener', () => {
    const handleJoinMission = jest.fn();
    handleJoinMission.mockReturnValue(store.missionsArray);
    handleJoinMission();
    render(
      <Provider store={store}>
        <table>
          <tbody>
            <tr>
              <Mission
                id={missionsArray[0].mission_id}
                name={missionsArray[0].mission_name}
                description={missionsArray[0].description}
                reserved={missionsArray[0].reserved}
              />
            </tr>
          </tbody>
        </table>
      </Provider>,
    );
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(handleJoinMission).toHaveBeenCalled();
  });
});
