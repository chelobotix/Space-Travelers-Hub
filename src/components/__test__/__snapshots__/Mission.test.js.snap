// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission Mission Tree 1`] = `
<table>
  <tbody>
    <tr>
      <td
        className="border title"
        data-testid="titleTD"
      >
        Mission test
      </td>
      <td
        className="border description"
      >
        Description Test
      </td>
      <td
        className="border status"
      >
        <span
          className="status1"
        >
          NOT A MEMBER
        </span>
      </td>
      <td
        className="border status"
      >
        <button
          className="btn1"
          onClick={[Function]}
          type="button"
        >
          Join Mission
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
